crates/neon/src/thread/mod.rs

```diff
@@ -0,0 +1,232 @@
+//!
+//! ### Why Not Use Standard TLS?
+//!
+//! Because the JavaScript engine may not tie JavaScript threads 1:1 to system threads,
+//! it is recommended to use this module instead of the Rust standard thread-local storage
+//! when tying data to a JavaScript thread.
+//!
+//! [environment]: https://nodejs.org/api/n-api.html#environment-life-cycle-apis
+//! [lifecycle]: https://raw.githubusercontent.com/neon-bindings/neon/main/doc/lifecycle.png
+//! [workers]: https://nodejs.org/api/worker_threads.html
+//! [threadId]: https://nodejs.org/api/worker_threads.html#workerthreadid
+
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::LocalCell;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A JavaScript thread-local container that owns its contents, similar to
+/// [`std::thread::LocalKey`](std::thread::LocalKey) but tied to a JavaScript thread rather
+/// than a system thread.
+///
+/// ### Initialization and Destruction
+///
+/// Initialization is dynamically performed on the first call to one of the `init` methods
+/// of `LocalKey`, and values that implement [`Drop`](std::ops::Drop) get destructed when
+/// the JavaScript thread exits, i.e. when a worker thread terminates or the main thread
+/// terminates on process exit.
+#[derive(Default)]
+pub struct LocalKey<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> LocalKey<T> {
+    /// Creates a new local value. This method is `const`, so it can be assigned to
+    /// static variables.
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send + 'static> LocalKey<T> {
+    /// Gets the current value of the cell. Returns `None` if the cell has not
+    /// yet been initialized.
+    pub fn get<'cx, 'a, C>(&self, cx: &'a mut C) -> Option<&'cx T>
+    where
+        C: Context<'cx>,
+    {
+        // Unwrap safety: The type bound LocalKey<T> and the fact that every LocalKey has a unique
+        // id guarantees that the cell is only ever assigned instances of type T.
+        let r: Option<&T> =
+            LocalCell::get(cx, self.id()).map(|value| value.downcast_ref().unwrap());
+
+        // Safety: Since the Box is immutable and heap-allocated, it's guaranteed not to
+        // move or change for the duration of the context.
+        unsafe { std::mem::transmute::<Option<&'a T>, Option<&'cx T>>(r) }
+    }
+
+    /// Gets the current value of the cell, initializing it with `value` if it has
+    /// not yet been initialized.
+    pub fn get_or_init<'cx, 'a, C>(&self, cx: &'a mut C, value: T) -> &'cx T
```

Comment:
Bike shedding a bit, but I would remove this method since it saves at most 8 characters. This frees up the name to rename `get_or_init_with` the shorter `get_or_init`.

```diff
- let global = GLOBAL.get_or_init(&mut cx, value);
+ let global = GLOBAL.get_or_init(&mut cx, move || value);
```

Reply:
Good idea, and this makes a better symmetry in the `get_or_init` / `get_or_try_init` naming. `once_cell` does the same thing.